plugins/wasmembed/secure-enclave/secure-enclave.cpp

```diff
@@ -0,0 +1,502 @@
+        auto len = strlen(val);
+        auto memIdxVar = wasmEngine->callRealloc(wasmName, {0, 0, 1, (int32_t)len});
+        auto memIdx = memIdxVar[0].i32();
+        auto mem = wasmEngine->getData(wasmName);
+        for (int i = 0; i < len; i++)
+        {
+            mem[memIdx + i] = val[i];
+        }
+        args.push_back(memIdx);
+        args.push_back((int32_t)len);
+    }
+    virtual void bindUTF8Param(const char *name, size32_t chars, const char *val)
+    {
+        TRACE("bindUTF8Param %s %d %s", name, chars, val);
+        auto memIdxVar = wasmEngine->callRealloc(wasmName, {0, 0, 1, (int32_t)chars});
+        auto memIdx = memIdxVar[0].i32();
+        auto mem = wasmEngine->getData(wasmName);
+        for (int i = 0; i < chars; i++)
+        {
+            mem[memIdx + i] = val[i];
+        }
+        args.push_back(memIdx);
+        args.push_back((int32_t)chars);
+    }
+    virtual void bindUnicodeParam(const char *name, size32_t chars, const UChar *val)
+    {
+        TRACE("bindUnicodeParam %s %d %S", name, chars, reinterpret_cast<const wchar_t *>(val));
+        auto memIdxVar = wasmEngine->callRealloc(wasmName, {0, 0, 2, (int32_t)chars * 2});
+        auto memIdx = memIdxVar[0].i32();
+        auto mem = wasmEngine->getData(wasmName);
+        for (int i = 0; i < chars * 2; i += 2)
+        {
+            mem[memIdx + i] = val[i];
+        }
+        args.push_back(memIdx);
+        args.push_back((int32_t)chars);
+    }
+    virtual void bindSetParam(const char *name, int elemType, size32_t elemSize, bool isAll, size32_t totalBytes, const void *setData)
+    {
+        TRACE("bindSetParam %s %d %d %d %d %p", name, elemType, elemSize, isAll, totalBytes, setData);
+    }
+    virtual void bindRowParam(const char *name, IOutputMetaData &metaVal, const byte *val) override
+    {
+        TRACE("bindRowParam %s %p", name, val);
+    }
+    virtual void bindDatasetParam(const char *name, IOutputMetaData &metaVal, IRowStream *val)
+    {
+        TRACE("bindDatasetParam %s %p", name, val);
+    }
+    virtual bool getBooleanResult()
+    {
+        TRACE("getBooleanResult");
+        return results[0].i32();
+    }
+    virtual void getDataResult(size32_t &__len, void *&__result)
+    {
+        TRACE("getDataResult");
+    }
+    virtual double getRealResult()
+    {
+        TRACE("getRealResult");
+        if (results[0].kind() == wasmtime::ValKind::F64)
+            return (int32_t)results[0].f64();
+        return results[0].f32();
+    }
+    virtual __int64 getSignedResult()
+    {
+        TRACE("getSignedResult");
+        if (results[0].kind() == wasmtime::ValKind::I64)
+            return (int32_t)results[0].i64();
+        return results[0].i32();
+    }
+    virtual unsigned __int64 getUnsignedResult()
+    {
+        TRACE("getUnsignedResult");
+        if (results[0].kind() == wasmtime::ValKind::I64)
+            return (int32_t)results[0].i64();
+        return results[0].i32();
```

Comment:
probably best to cast to unsigned first - otherwise it is not clear if it is signed extended